app/services/update_total_price.rb

```diff
@@ -1,23 +1,34 @@
 class UpdateTotalPrice
-  attr_reader :order, :price, :quantity, :action
+  attr_reader :order, :order_item, :quantity, :action, :delivery, :coupon
 
-  def initialize(order, order_item, action)
+  def initialize(order, order_item: nil, action: nil, delivery: nil, coupon: nil)
     @order = order
-    @price = order_item.price
-    @quantity = order_item.quantity
+    @order_item = order_item
     @action = action
+    @delivery = delivery
+    @coupon = coupon
   end
 
   def self.call(*args)
     new(*args).call
   end
 
   def call
-    difference = price * quantity - order.discount(price)
+    if delivery
+      order.increment!(:total_price, @delivery.price)
+    end
+
+    if coupon
+      order.decrement!(:total_price, @order.discount)
+    end
+
+    if order_item && action
+      difference = order_item.price * order_item.quantity - order.discount(order_item.price)
 
-    case action
-    when :remove then order.decrement!(:total_price, difference)
-    when :add then order.increment!(:total_price, difference)
+      case action
+      when :remove then order.decrement!(:total_price, difference)
+      when :add then order.increment!(:total_price, difference)
+      end
```

Comment:
Why Rubocop pass this method? It is too complicated.

It should be something like 
```
def call
  return order.increment!(:total_price, difference) if magic_happens?
  order.decrement!(:total_price, difference)
end

def difference
 ...
end

def magic_happens?
 ...
end
```